S2t new config

okuchaiev opened the pull request:
* Introduces ModelPTConfig as a base for all ModelPT configuration
* Introduces and changes EncDecCTCModel to be instantiated with EncDecCTCModelConfig object
* Adjusts config and speech_to_text.py accordingly

titu1994 commented on examples/asr/speech_to_text.py:
```diff
@@ -38,62 +39,36 @@
 
 
-@hydra.main(config_path="conf", config_name="config")
+@hydra.main(config_path="conf", config_name="config", strict=None)
```
`strict=None` is default value, can be removed.

okuchaiev replied:
removed

titu1994 commented on examples/asr/conf/config.yaml:
```diff
@@ -60,122 +16,164 @@ AudioToTextDataLayer:
+    name: novograd
+    # cls: nemo.core.optim.optimizers.Novograd
+    lr: .01
+    trainer: null
```
`trainer` should not be here, optim instantiates objects directly from class, so please adjust as shown below

titu1994 commented on examples/asr/speech_to_text.py:
```diff
@@ -38,62 +39,36 @@
+@hydra.main(config_path="conf", config_name="config", strict=None)
 def main(cfg):
+    # omegaconf merg trainer stuff to optim - this is necessary to be able to correctly setup LR scheduler
+    cfg.model.optim.trainer = cfg.pl.trainer
```
This should be `cfg.model.pl = cfg.pl`, please look below how we can use it then.

okuchaiev replied:
ok

titu1994 commented on nemo/collections/asr/models/ctc_models.py:
```diff
@@ -68,22 +93,39 @@ def __setup_dataloader_from_config(config: Optional[Dict]):
-        self.__scheduler = prepare_lr_scheduler(
-            optimizer=self.__optimizer, scheduler_config=optim_config, train_dataloader=self.__train_dl
+        # Setup optimizer and scheduler
+        if 'sched' in optim_config and 'trainer' in optim_config:
```
`'trainer' in self._cfg.model.pl`

okuchaiev replied:
ok

okuchaiev replied:
but without . model since self._cfg is already model's config

titu1994 commented on nemo/collections/asr/models/ctc_models.py:
```diff
@@ -68,22 +93,39 @@ def __setup_dataloader_from_config(config: Optional[Dict]):
-            optimizer=self.__optimizer, scheduler_config=optim_config, train_dataloader=self.__train_dl
+        # Setup optimizer and scheduler
+        if 'sched' in optim_config and 'trainer' in optim_config:
+            if optim_config.trainer.max_steps is None:
```
`if self._cfg.model.pl.trainer.max_steps is None:`

okuchaiev replied:
but without . model since self._cfg is already model's config

titu1994 commented on nemo/collections/asr/models/ctc_models.py:
```diff
@@ -68,22 +93,39 @@ def __setup_dataloader_from_config(config: Optional[Dict]):
+        # Setup optimizer and scheduler
+        if 'sched' in optim_config and 'trainer' in optim_config:
+            if optim_config.trainer.max_steps is None:
+                if optim_config.trainer.gpus == 0:
```
`if self._cfg.model.pl.trainer.gpus == 0`

okuchaiev replied:
but without . model since self._cfg is already model's config

titu1994 commented on nemo/collections/asr/models/ctc_models.py:
```diff
@@ -68,22 +93,39 @@ def __setup_dataloader_from_config(config: Optional[Dict]):
+                    # training on CPU
+                    iters_per_batch = optim_config.trainer.max_epochs / float(
+                        optim_config.trainer.num_nodes * optim_config.trainer.accumulate_grad_batches
+                    )
```
```python
iters_per_batch = self._cfg.model.pl.trainer.max_epochs / float(
                        self._cfg.model.pl.trainer.num_nodes * self._cfg.model.pl.trainer.accumulate_grad_batches
                    )

```

okuchaiev replied:
but without . model since self._cfg is already model's config

titu1994 commented on nemo/collections/asr/models/ctc_models.py:
```diff
@@ -68,22 +93,39 @@ def __setup_dataloader_from_config(config: Optional[Dict]):
+                        optim_config.trainer.gpus
+                        * optim_config.trainer.num_nodes
+                        * optim_config.trainer.accumulate_grad_batches
+                    )
```
```python
iters_per_batch = self._cfg.model.pl.trainer.max_epochs / float(
                        self._cfg.model.pl.trainer.gpus
                        * self._cfg.model.pl.trainer.num_nodes
                        * self._cfg.model.pl.trainer.accumulate_grad_batches
                    )
```

okuchaiev replied:
but without . model since self._cfg is already model's config

titu1994 commented on nemo/core/classes/modelPT.py:
```diff
@@ -23,16 +24,39 @@
     """
 
+    def __init__(self, cfg: ModelPTConfig = None):
+        super().__init__()
```
Please attach self._cfg = cfg here after super().***

okuchaiev replied:
done

titu1994 commented on nemo/core/classes/modelPT.py:
```diff
@@ -23,16 +24,39 @@
+        self._scheduler = None
+
+        if cfg is not None:
+            self.setup_training_data(cfg.train_ds)
```
Maybe lets do a conditional check here to see if `cfg.train_ds` is None or not? Same for validation_ds below. Also, maybe test and add test_ds here too?

okuchaiev replied:
done

titu1994 commented on nemo/core/classes/modelPT.py:
```diff
@@ -155,3 +179,15 @@ def setup_optimization(self, optim_config: Optional[Union[DictConfig, Dict]] = N
+    def train_dataloader(self):
+        return self._train_dl
+
+    def val_dataloader(self):
```
Maybe setup `test_dataloader(self)` as well?

okuchaiev replied:
added

okuchaiev replied:
actually there is some PTL strangeness with this, so not doing it for now

titu1994 replied:
Sounds good.

okuchaiev commented:
> Some minor comments regarding refactoring of the config object for Optimizers. We can probably push it all the way down into the `super().setup_optimization()` call in the beginning at this point so it is completely hidden from both users and our team as well.

moved all setup_optimizatio() logic into ModelPT clas

titu1994 commented on nemo/core/classes/modelPT.py:
```diff
@@ -178,7 +202,12 @@ def setup_optimization(self, optim_config: Optional[Union[DictConfig, Dict]] = N
+            self._optimizer = optimizer
+            # return optimizer
+
+        self._scheduler = prepare_lr_scheduler(
```
There are a few more return optimizers above inside the function, so please also add this lines there.

okuchaiev replied:
fixed. thanks for catching that!

okuchaiev commented:
> There's a few extra return steps for optimizer, so we need to instantiate the schedulers there too. All else looks great

- fixed them. And adjusted NLP models to the new base class so that CI passes now
